### PR TITLE
feature: add default arguments and fix compiling issues in ALCF Polaris

### PR DESCRIPTION
**Pull request by rayandrew**
This added feature to support capturing `step` default params while annotating `func(step=2)`

All tests are passed.

Along with this pull request, I added `developer-guide` on how to develop this on ALCF Polaris machine.
Turns out, it is quite tricky, because `brahma` is being interfered by `darshan` POSIX interceptor and causing `segfault`

Disabling it giving me error free while running all tests.

### CODE REVIEW

**Comment by rayandrew** [comment]
Just FYI,

This is the stacktrace if darshan module is enabled

```
#0  0x00007f866c3aed42 in __parse_one_specmb () from /lib64/libc.so.6
Missing separate debuginfos, use: zypper install glibc-debuginfo-2.31-150300.63.1.x86_64 krb5-debuginfo-1.20.1-150500.3.3.1.x86_64 libatomic1-debuginfo-13.2.1+git7813-150000.1.6.1.x86_64 libbrotlicommon1-debuginfo-1.0.7-3.3.1.x86_64 libbrotlidec1-debuginfo-1.0.7-3.3.1.x86_64 libcom_err2-debuginfo-1.46.4-150400.3.3.1.x86_64 libcurl4-debuginfo-8.0.1-150400.5.41.1.x86_64 libgcc_s1-debuginfo-13.2.1+git7813-150000.1.6.1.x86_64 libgfortran5-debuginfo-13.2.1+git7813-150000.1.6.1.x86_64 libidn2-0-debuginfo-2.2.0-3.6.1.x86_64 libjansson4-debuginfo-2.14-150000.3.5.1.x86_64 libjitterentropy3-debuginfo-3.4.0-150000.1.9.1.x86_64 libjson-c3-debuginfo-0.13-3.3.1.x86_64 libkeyutils1-debuginfo-1.6.3-5.6.1.x86_64 libldap-2_4-2-debuginfo-2.4.46-150200.14.17.1.x86_64 libnghttp2-14-debuginfo-1.40.0-150200.12.1.x86_64 libnl3-200-debuginfo-3.3.0-1.29.x86_64 libopenssl1_1-debuginfo-1.1.1l-150500.17.25.1.x86_64 libpcre1-debuginfo-8.45-150000.20.13.1.x86_64 libpsl5-debuginfo-0.20.1-150000.3.3.1.x86_64 libquadmath0-debuginfo-13.2.1+git7813-150000.1.6.1.x86_64 libsasl2-3-debuginfo-2.1.28-150500.1.1.x86_64 libselinux1-debuginfo-3.1-150400.1.69.x86_64 libstdc++6-debuginfo-13.2.1+git7813-150000.1.6.1.x86_64 libunistring2-debuginfo-0.9.10-1.1.x86_64 libz1-debuginfo-1.2.13-150500.4.3.1.x86_64 libzstd1-debuginfo-1.5.0-150400.3.3.1.x86_64
(gdb) bt full
#0  0x00007f866c3aed42 in __parse_one_specmb () from /lib64/libc.so.6
No symbol table info available.
#1  0x00007f866c3a2f39 in printf_positional () from /lib64/libc.so.6
No symbol table info available.
#2  0x00007f866c3a728d in __vfprintf_internal () from /lib64/libc.so.6
No symbol table info available.
#3  0x00007f866c3b3851 in __vsprintf_internal () from /lib64/libc.so.6
No symbol table info available.
#4  0x00007f866cb185af in cpplogger::Logger::log(cpplogger::LoggerType, char const*, ...) [clone .constprop.0] ()
   from /lus/eagle/projects/MDClimSim/rayandrew/dftracer/build/temp.linux-x86_64-cpython-311/dftracer.pydftracer/lib64/libdftracer.so.3.0.0
No symbol table info available.
#5  0x00007f866cb1d3e1 in brahma::STDIODFTracer::fwrite(void const*, unsigned long, unsigned long, _IO_FILE*) ()
   from /lus/eagle/projects/MDClimSim/rayandrew/dftracer/build/temp.linux-x86_64-cpython-311/dftracer.pydftracer/lib64/libdftracer.so.3.0.0
No symbol table info available.
#6  0x00007f866cac8432 in fwrite_wrapper(void const*, unsigned long, unsigned long, _IO_FILE*) ()
   from /lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/.venvs/polaris/2024-04-29/lib/python3.11/site-packages/dftracer/lib64/libbrahma.so.1
No symbol table info available.
#7  0x00007f866c93a274 in fwrite (ptr=0x7ffcb5195230, size=1, nmemb=200, stream=<optimized out>) at darshan-stdio.c:461
        ret = <optimized out>
        tm1 = 0
        tm2 = <optimized out>
        __darshan_disabled = 1
#8  0x00007f866cac8432 in fwrite_wrapper(void const*, unsigned long, unsigned long, _IO_FILE*) ()
   from /lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/.venvs/polaris/2024-04-29/lib/python3.11/site-packages/dftracer/lib64/libbrahma.so.1
No symbol table info available.
#9  0x00007f866c93a274 in fwrite (ptr=0x7ffcb5195230, size=1, nmemb=200, stream=<optimized out>) at darshan-stdio.c:461
        ret = <optimized out>
        tm1 = 0
        tm2 = <optimized out>
        __darshan_disabled = 1
#10 0x00007f866cac8432 in fwrite_wrapper(void const*, unsigned long, unsigned long, _IO_FILE*) ()
   from /lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/.venvs/polaris/2024-04-29/lib/python3.11/site-packages/dftracer/lib64/libbrahma.so.1
No symbol table info available.
#11 0x00007f866c93a274 in fwrite (ptr=0x7ffcb5195230, size=1, nmemb=200, stream=<optimized out>) at darshan-stdio.c:461
        ret = <optimized out>
        tm1 = 0
        tm2 = <optimized out>
        __darshan_disabled = 1
#12 0x00007f866cac8432 in fwrite_wrapper(void const*, unsigned long, unsigned long, _IO_FILE*) ()
   from /lus/eagle/projects/MDClimSim/rayandrew/stormer-dev-pure/.venvs/polaris/2024-04-29/lib/python3.11/site-packages/dftracer/lib64/libbrahma.so.1
No symbol table info available.
#13 0x00007f866c93a274 in fwrite (ptr=0x7ffcb5195230, size=1, nmemb=200, stream=<optimized out>) at darshan-stdio.c:461
        ret = <optimized out>
        tm1 = 0
...
```

**Comment by hariharan-devarajan** [comment]
> giving me error free while running all

Yeah darshan does static linking which will significantly screw up everything. Brahma can be stacked for multiple tools correctly but not darshan. Darshan folks are currently working on fixing this and using GOTCHA.

**Comment by rayandrew** [comment]
> Yeah darshan does static linking which will significantly screw up everything. Brahma can be stacked for multiple tools correctly but not darshan. Darshan folks are currently working on fixing this and using GOTCHA.

Ah good to know. Perhaps a lot of unknown errors that I encountered a few weeks ago probably caused by Darshan.

not conclusive though since the second profiler job was running fine when I retried it.

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -22,7 +22,7 @@ add_dependencies(test_c ${PROJECT_NAME}_preload)
 function(set_common_properties test_name)
     set_property(TEST ${test_name} APPEND PROPERTY ENVIRONMENT DFTRACER_LOG_LEVEL=DEBUG)
     set_property(TEST ${test_name} APPEND PROPERTY ENVIRONMENT DFTRACER_TRACE_COMPRESSION=0)
-    set_property(TEST ${test_name} APPEND PROPERTY ENVIRONMENT LD_LIBRARY_PATH=${CMAKE_BINARY_DIR}/${DFTRACER_LIBDIR})
+    set_property(TEST ${test_name} APPEND PROPERTY ENVIRONMENT LD_LIBRARY_PATH=${CMAKE_BINARY_DIR}/${DFTRACER_LIBDIR}:$ENV{TEST_LD_LIBRARY_PATH})
```
Comment (rayandrew) [inline test/CMakeLists.txt:25]: Adding this (along with the line 128) to support additional TEST_LD_LIBRARY_PATH.

In ALCF Polaris, sometimes we have problem where we cannot find `libfabric.so` (our last meeting showed the problem)

File: setup.py [hunk 2]
```diff
@@ -194,7 +194,7 @@ def build_extension(self, ext: CMakeExtension) -> None:
                     "dfanalyzer": [
                         "bokeh>=2.4.2",
                         "pybind11",
-                        "zindex_py==0.0.1",
+                        "zindex_py@git+https://github.com/hariharan-devarajan/zindex.git",
```
Comment (rayandrew) [inline setup.py:197]: This is also to support the new version of zindex from git where I encounterd `-lhwloc` issue on Polaris. let me know whether you want to push new version of zindex or just use the git version

Comment (hariharan-devarajan) [reply setup.py:197]: I just release 0.0.2 for zindex on pypi. Use zindex_py==0.0.2

File: test/CMakeLists.txt (hunk 1 above)
Comment (hariharan-devarajan) [reply test/CMakeLists.txt:25]: U should not add it like this. Add a option on CMake and then use it internally. Thats is less opaque as other people can use it correctly.

File: test/CMakeLists.txt [hunk 3]
```diff
@@ -125,7 +125,7 @@ set(test_name test_py_both)
 df_add_test(${test_name} ${DFTRACER_PYTHON_EXE} ${CMAKE_CURRENT_SOURCE_DIR}/py/test.py --format=npz --data_dir=${CMAKE_CURRENT_BINARY_DIR}/data)
 set_common_properties(${test_name})
 set_property(TEST ${test_name} APPEND PROPERTY ENVIRONMENT PYTHONPATH=$ENV{PYTHONPATH}:${CMAKE_SOURCE_DIR}/venv/${DFTRACER_LIBDIR})
-set_property(TEST ${test_name} APPEND PROPERTY ENVIRONMENT LD_LIBRARY_PATH=${CMAKE_BINARY_DIR}/${DFTRACER_LIBDIR}:${CMAKE_SOURCE_DIR}/dependency/.spack-env/view/lib64)
+set_property(TEST ${test_name} APPEND PROPERTY ENVIRONMENT LD_LIBRARY_PATH=${CMAKE_BINARY_DIR}/${DFTRACER_LIBDIR}:${CMAKE_SOURCE_DIR}/dependency/.spack-env/view/lib64:$ENV{TEST_LD_LIBRARY_PATH})
```
Comment (hariharan-devarajan) [inline test/CMakeLists.txt:128]: Lets add TEST_LD_LIBRARY_PATH to CMake Options.

**Review by hariharan-devarajan — CHANGES_REQUESTED** [review]
Some minor fixes

**Comment by hariharan-devarajan** [comment]
> > Yeah darshan does static linking which will significantly screw up everything. Brahma can be stacked for multiple tools correctly but not darshan. Darshan folks are currently working on fixing this and using GOTCHA.
> 
> Ah good to know. Perhaps a lot of unknown errors that I encountered a few weeks ago probably caused by Darshan.
> 
> not conclusive though since the second profiler job was running fine when I retried it.

Could u install DFTracer with DEBUG symbols on. do 

`export DFTRACER_BUILD_TYPE=Debug`
`pip install --force-reinstall dftracer`

This will show line numbers for us to investigate the issue.

**Comment by rayandrew** [comment]
> Could u install DFTracer with DEBUG symbols on. do
> 
> `export DFTRACER_BUILD_TYPE=Debug` `pip install --force-reinstall pydftracer`
> 
> This will show line numbers for us to investigate the issue.

Yes, next time I run the jobs, I will use the debug version of dftracer

File: test/CMakeLists.txt (hunk 1 above)
Comment (rayandrew) [reply test/CMakeLists.txt]: fixed using `DFTRACER_TEST_LD_LIBRARY_PATH` options on cmake and modified setup.py as well

File: setup.py (hunk 2 above)
Comment (rayandrew) [reply setup.py]: fixed

File: test/CMakeLists.txt (hunk 3 above)
Comment (rayandrew) [reply test/CMakeLists.txt]: fixed using `DFTRACER_TEST_LD_LIBRARY_PATH`

**Comment by rayandrew** [comment]
fixed minor issues above, added test case and added the cmake options (along with its usage in `setup.py`)

**Review by hariharan-devarajan — APPROVED** [review]
This looks good.

**Comment by rayandrew** [comment]
thanks @hariharan-devarajan you can merge it, I am done with this branch

**Pull request merged**